PRCS/prc-2.md

```diff
@@ -0,0 +1,157 @@
+    /// @dev This function should be protected with appropriate access control mechanisms.
+    /// The new value should be checked against a sane upper limit constant, which if exceeded,
+    /// should cause a revert.
+    /// Emits `LockTimeUpdate` event.
+    /// @param newLockTime New lockTime value
+    function setLockTime(uint256 newLockTime) external;
+}
+```
+
+A Hololocker implementation MUST implement the IERC721Receiver interface to be able to receive ERC721 assets via `IERC721.safeTransferFrom`.
+It MUST initialize a lock similarly as in the `lock` function, and it MUST emit the `Lock` event
+
+```solidity
+interface IERC721Receiver {
+    /// @dev Whenever an {IERC721} `tokenId` token is transferred to this contract via {IERC721-safeTransferFrom}
+    ///  by `operator` from `from`, this function is called.
+    ///  It must return its Solidity selector to confirm the token transfer.
+    ///  If any other value is returned or the interface is not implemented by the recipient, the transfer will be
+    ///  reverted.
+    ///  The selector can be obtained in Solidity with `IERC721Receiver.onERC721Received.selector`.
+    ///  Note: the contract address is always the message sender.
+    /// @param _operator The address which called `safeTransferFrom` function
+    /// @param _from The address which previously owned the token
+    /// @param _tokenId The NFT identifier which is being transferred
+    /// @param _data Additional data with no specified format
+    /// @return `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`
+    ///  unless throwing
+    function onERC721Received(address _operator, address _from, uint256 _tokenId, bytes _data) external returns(bytes4);
+}
+```
+
+## Rationale
+
+The Hololocker contract is in its core a NFT staking contract. While there have been attempts at EIPs on this matter, such as [https://eips.ethereum.org/EIPS/eip-4987](EIP-4987), none gained enough traction and approval to be finalized.
+This proposal therefore proposes a standard for light-weight implementation of such mechanism.
+
+The main functions accept an array of token addresses and token IDs to be able to accomodate manipulations with multiple NFTs in one transaction, to save on transaction costs.
+
+### NFT locking UX
+
+Upon locking an NFT in an application that integrates the Hololocker, it might happen that users will get confused as to where did their NFTs go. An option to mint a sort of a "receipt NFT" in exchange for the locked NFT was thought of, but it does not sufficiently solve the inconvenience and would greatly increase the complexity and transactional costs. Therefore, this approach is not recommended.
+
+### Projecting other standards (ERC20, ERC1155...)
+
+Should there be a need for supporting other token standards than ERC721, the interface should be appropriately modified to accomodate for the differences between the standards' methods of token identification.

```

Comment:
I would expect more detail mentioning the constant registry address thanks to using a proxy and how we use a to have a deterministic address for each different standard type

Reply:
But that's purely an implementation detail, I don't think it belongs into a standard describing an interface 😕 
It is indeed described in the implementation repo, linked in the `Reference Implementation` section

Reply:
Part of the standard is knowing the standard address you should be looking at. It's a constant across all chains, so it's worth mentioning imo

Reply:
Fair enough, will add it.